Use structured find/replace to avoid false positives

glyn opened the pull request:
A side-effect of structured find/replace is that modified YAML is reformatted (using an indentation of 2 characters).

There is now no distinction between "single" and "multiple" user defined images. A JSON path may identify zero or more nodes of a YAML document that contain images.

In order to support larger YAML files, it was necessary to obtain a fix (or at least a workaround) to a Kubernetes issue. See TODOs in this PR and #76 for details of some code that was forked from Kubernetes.

With structured find/replace, the config add-image command, which relies on global find/replace, is no longer supported.

Also, made some improvements to linting to avoid false positives and other issues:
* Install stable version of golangci-lint (See https://github.com/golangci/golangci-lint#binary)
* Exclude `bodyclose` linter to avoid false positives (worse than that, the linter seems to pass or fail non-deterministically)
* Make sure that only the specified linters run, by disabling the test

Fixes https://github.com/bryanl/sheaf/issues/30
Fixes https://github.com/bryanl/sheaf/issues/69

glyn commented:
@bryanl This PR is still a little rough round the edges, but is starting to come together. the image replacement logic now uses a structured find/replace to avoid false positives. With this new approach, I don’t think we need to distinguish between single and multi UDIs.

It’s still a draft PR because there are some TODOs and FIXMEs that need addressing. I haven’t tried it out in anger (i.e. on the command line), but it _should_ be able to cope. Oh and I need to copy across the documentation of the JSON path syntax and semantics.

In due course, I’d also like to extend the UDI syntax so that you can use a wildcard to match all API version and/or all kinds. Then we can express the built-in image paths similarly to a UDI and possibly common up some code.

I'm not sure why the CI lint job has started passing, but at least that's now consistent with running it locally.

I'll get back to this on Monday, so any comments before then gratefully received.

bryanl commented:
@glyn rebase against master and re-push. I'm ready to test this out.

glyn commented:
> @glyn rebase against master and re-push. I'm ready to test this out.

@bryanl Done. I haven't tested the commands yet, so if you find anything unexpected, please let me know rather than wasting much time on it.

bryanl commented:
Thanks. Will give it a run through.

glyn commented:
The distinction between single and multi UDIs has now been removed. I also added JSON path syntax and semantics to the set-udi command.

bryanl commented:
@glyn Could you update the readme with the new syntax?

bryanl commented on internal/yamlutil/decoder.go:
```diff
@@ -0,0 +1,351 @@
+/*
```
This is more of a nit than anything else, but could you move this code to a designated 3rd party namespace to make it easier to see that this is upstream code?

glyn replied:
Fixed. I think this was more than a nit.

glyn commented:
@bryanl:

> @glyn Could you update the readme with the new syntax?

Done.

glyn commented:
@bryanl Please note that the `config add-image` command is deleted since it makes no sense without global find/replace.

glyn commented:
@bryanl I addressed your comments and deleted the config set-image command as mentioned above. The coverage has decreased marginally, presumably because deleting some well covered code affects the overall stats. PTAL and merge if you're happy.

glyn commented:
@bryanl is happy to merge.
